File: web/server/src/db/models/issue.js
const { Model, DataTypes } = require('sequelize');

module.exports = class Issue extends Model {
  static init(sequelize) {
    return super.init(
      {
        num: {
          type: DataTypes.INTEGER.UNSIGNED,
          autoIncrement: true,
          primaryKey: true,
        },
        title: {
          type: DataTypes.STRING,
          allowNull: false,
        },
        isClosed: {
          type: DataTypes.BOOLEAN,
          allowNull: false,
          defaultValue: false,
        },
        isDeleted: {
          type: DataTypes.BOOLEAN,
          allowNull: false,
          defaultValue: false,
        },
      },
      {
        sequelize,
        underscored: true,
        modelName: 'Issue',
        tableName: 'issues',
        charset: 'utf8',
        collate: 'utf8_general_ci',
      },
    );
  }
  static associate({ Issue, Comment, Milestone, User, Label }) {
    Issue.hasMany(Comment, {
      foreignKey: {
        name: 'issue_num',
        allowNull: false,
      },
      sourceKey: 'num',
      as: 'comments',
    });
    Issue.belongsTo(Milestone, {
      foreignKey: 'milestone_num',
      targetKey: 'num',
    });
    Issue.belongsTo(User, {
      foreignKey: {
        name: 'user_num',
        allowNull: false,
      },
      targetKey: 'num',
      as: 'author',
    });
    Issue.belongsToMany(User, {
      foreignKey: 'issue_num',
      through: 'assignments',
      as: 'assignees',
      timestamps: false,
    });
    Issue.belongsToMany(Label, {
      foreignKey: 'issue_num',
      through: 'labelings',
      as: 'labels',
      timestamps: false,
    });
  }
};

File: iOS/IssueTracker/IssueTracker/IssueListScene/Models/Issue.swift
//
//  Issue.swift
//  IssueTracker
//
//  Created by 홍경표 on 2020/11/02.
//

import Foundation

struct IssueAPI: Codable {
    var open: Int
    var closed: Int
    var issues: [Issue]
}

struct Issue {
    var id: Int
    var title: String
    var author: User?
    var createdAt: String
    var isClosed: Bool
    var isDeleted: Bool?
    var labels: [Label]
    var milestone: Milestone?
    var assignees: [User]
    var comment: Comment?
    var comments: [Comment]?
    
    /**
    새로 생성을 위한 initializer
     - id: -1
     - 나머지는 모두 nil or []
     */
    init(
        id: Int = -1,
        title: String,
        author: User? = nil,
        createdAt: String = "",
        isClosed: Bool = false,
        isDeleted: Bool = false,
        labels: [Label] = [],
        milestone: Milestone? = nil,
        assignees: [User] = [],
        comment: Comment? = nil,
        comments: [Comment]? = nil
    ) {
        self.id = id
        self.title = title
        self.author = author
        self.createdAt = createdAt
        self.isClosed = isClosed
        self.isDeleted = isDeleted
        self.labels = labels
        self.milestone = milestone
        self.assignees = assignees
        self.comment = comment
        self.comments = comments
    }
    
    var createData: [String: Any] {
        return [
            "title": title,
            "content": comment?.content ?? "",
            "labels": labels.compactMap { $0.id },
            "assignees": assignees.compactMap { $0.num },
            "milestoneNum": milestone ?? NSNull()
        ]
    }
    
    var titleData: [String: Any] {
        return [
            "title": title
        ]
    }
    
    var milestoneData: [String: Any] {
        return [
            "milestoneNum": milestone?.id ?? NSNull()
        ]
    }
    
    var statusData: [String: Any] {
        return [
            "isClosed": isClosed
        ]
    }
}

extension Issue: Codable {
    enum CodingKeys: String, CodingKey {
        case id = "num"
        case milestone = "Milestone"
        case title, author, createdAt, isClosed, isDeleted, labels, assignees, comment, comments
    }
}
